treest.1
.TH TREEST 1 treest-TREEST_VERSION

.SH NAME
treest \- stateful tree view

.SH SYNOPSIS
.B treest
[--printer=\fINAME\fR] [--\fILONGOPTIONS\fR] [-\fIFLAGS\fR]
[[--] \fIROOT\fR]

.SH DESCRIPTION
.PP
Print a tree-like representation of a directory to standard
output.  Take interactive commands from standard input.
When supported and standard output is a TTY, it is set to
raw mode.

.SH OPTIONS
.TP
\fB\-\-help
Show the quick help.
.TP
\fB\-\-version
Print the version and build features successfully.
.TP
\fB\-\-printer=\fINAME\fR
Change for the specified printer.
.TP
\fB\-\-ignore=\fISHELLPAT\fR
Add the given shell pattern to the ignore list.  See also
\fBfnmatch\fR(3).
.TP
\fB\-\-rcfile=\fIFILE\fR
Once ready but before reading from stdin, read from the
provided file first.
.TP
\fB\-\-\fILONGOPTIONS\fR
Other options are forwarded to the last specified printer.

.SH FLAGS
.PP
These are the common default flags.  Although they can be
overriden by the selected printer, they usually keep the
same meaning.
.TS
tab (@);
l lx.
-A@T{
do not ignore entries starting with . (except . and ..)
T}
-B@T{
do not list entries ending with ~
T}
-I@T{
toggle the filtering by ignored patterns
T}
-S@T{
sort by size
T}
-X@T{
sort by extension
T}
-c@T{
sort by ctime
T}
-d@T{
sort directories first
T}
-r@T{
reverse sorting order
T}
-t@T{
sort by mtime
T}
-u@T{
sort by atime
T}
-w@T{
toggle watch mode: watches are set on every unfolded
directories (see the \fBinotify(7)\fR)
T}
.TE

.SH COMMANDS
.PP
The following commands are read from the user input.
Note that from any prompt or pending state, \fB'^['\fR
(\fBESC\fR) should bring it back to "normal mode" (except
when using \fBreadline\fR(3) with a tty true user input).
.TS
tab (@);
l lx.
^C@T{
quit
T}
^H@T{
toggle the ignore global flag
T}
^L@T{
refresh the view
T}
^N@T{
go to the next visible node
T}
^P@T{
go to the previous visible node
T}
^R@T{
reload the directory at the cursor
T}
^Z@T{
suspend
T}
!@T{
execute a shell command, use {} for the full path at the cursor
T}
"@T{
fill or empty a register, use `?"` to see used registers content
T}
#@T{
(comment) ignore input until the end of line
T}
$@T{
find the next node which name ends with
T}
(@T{
run commands if
T}
)@T{
run commands ifnot
T}
-@T{
toggle a flag
T}
\.@T{
re-run the last command
T}
/@T{
find the next node which name contains
T}
:@T{
execute a printer command
T}
;@T{
refresh the view
T}
\= @T{
fold recursively at the cursor
T}
?@T{
print help for a given command
T}
C@T{
fold at the given path
T}
H@T{
fold at the cursor
T}
L@T{
unfold at the cursor
T}
N@T{
continue search backward
T}
O@T{
unfold at the given path
T}
Q@T{
quit with an exit code (by default indicating failure)
T}
[@T{
go to the parent's first child
T}
\\@T{
run a register as a sequence of commands
T}
]@T{
go to the parent's last child
T}
^@T{
find the next node which name starts with
T}
`@T{
go to the root
T}
c@T{
go to and fold at the given path
T}
h@T{
go to the parent directory
T}
j@T{
go to the next node
T}
k@T{
go to the previous node
T}
l@T{
go to the directory's first child (unfold if needed)
T}
n@T{
continue search forward
T}
o@T{
go to and unfold at the given path
T}
q@T{
quit
T}
{@T{
run commands while
T}
|@T{
pipe content into a shell command, use {} for the full path at the cursor
T}
}@T{
run commands whilenot
T}
~@T{
reload at the root (read the whole tree from file system)
T}
.TE

.SH PRINTERS
.PP
The flags available depend on the printer used.  Unknown
flags are silently ignored.  The default printer is
"\fBascii\fR".

.TP
\fBascii\fR - simple default and debug printer
.TS
tab (@);
l lx.
-F@T{
classify (append as decorator one of */=@|)
T}
-P@T{
show the node's relative path instead of only the name
T}
-h@T{
with -s, show sizes in a more human readable way
T}
-i@T{
show child count for directories and node index for child
T}
-l@T{
when the directory is from a link, show addresses
T}
-s@T{
show the size of files
T}
.TE

.TP
\fBfancy\fR - standard terminal printer
If it was built with \fBgit2\fR(3), it will also use the
appropriate ignore rules for given root directory.
.TS
tab (@);
l lx.
-C@T{
colorize the output
T}
-F@T{
classify (append as decorator one of */=@|)
T}
-T@T{
do not send terminal escape sequences (hiding cursor and
alternative screen buffer)
T}
-h@T{
with -s, show sizes in a more human readable way
T}
-j@T{
when a directory has a single child, show on the same line
T}
-q@T{
redraw after leaving the alternative buffer screen (ie.
'keep the tree when leaving')
T}
-s@T{
show the size of files
T}

^E@T{
forward one line
T}
^Y@T{
backward one line
T}
^D@T{
forward one half-window
T}
^U@T{
backward one half-window
T}
^F@T{
forward one window
T}
^B@T{
backward one window
T}
.TE

.SH BUGS
The "fancy" printer does not scroll to keep the cursor
on-screen.  It also does not update its internal window
size without an explicit redraw.

.SH NOTES
The shell (!) and pipe (|) commands use the
\fBsystem\fR(3) if a shell is indeed available.  The
command is executed from the cwd of treest, which is
not changed from its own invocation.

.SH "SEE ALSO"
.BR tree (1)

.SH AUTHOR
Grenier Celestin


treest.c
#include "./treest.h"
#include "./commands.h"

char oups[_MAX_PATH+14];
char* prog;
char cwd[_MAX_PATH];
bool is_tty;
bool is_raw;
struct GFlags gflags;
struct Node root, * cursor;
struct Printer* selected_printer;

static int LOOPBACK_FILENO[2] = {-1, -1};
#define LB_READ 0
#define LB_WRITE 1
static int NOTIFY_FILENO = -1;
static struct WatchAssoc {
    int wd;
    struct Node* node;
    struct WatchAssoc* next;
}* watch_assocs;
static fd_set user_fds;

struct Node* node_alloc(struct Node* parent, char* path) {
    struct Node fill = {
        .path=path,
        .name=strrchr(path, '/')+1,
        .parent=parent,
    };

    if (lstat(path, &fill.stat) < 0) return NULL;
    fill.type = S_IFMT & fill.stat.st_mode;

    struct Node* niw; may_malloc(niw, sizeof(struct Node));
    memcpy(niw, &fill, sizeof(struct Node));

    switch (fill.type) {
        case Type_REG:
            if (fill.stat.st_mode & S_IXUSR)
                niw->type = Type_EXEC;
            break;

        case Type_LNK:
            lnk_resolve(niw);
            break;

        default: ;
    }

    return niw;
}

void node_free(struct Node* node) {
    switch (node->type) {
        case Type_DIR: dir_free(node); break;
        case Type_LNK: lnk_free(node); break;
        default: ;
    }
    free(node->path);
    node->path = NULL;
    node->name = NULL;
    if (node == cursor)
        cursor = node->parent
            ? node->parent
            : &root;
}

void dir_free(struct Node* node) {
    for (size_t k = 0; k < node->count; k++) {
        node_free(node->as.dir.children[k]);
        free(node->as.dir.children[k]);
        node->as.dir.children[k] = NULL;
    }
    node->count = 0;

    if (gflags.watch && watch_assocs) {
        struct WatchAssoc* wa;
        struct WatchAssoc* pwa = watch_assocs;
        do {
            wa = pwa->next;
            if (wa && wa->node == node) break;
        } while ((pwa = wa));
        if (wa) {
            inotify_rm_watch(NOTIFY_FILENO, wa->wd);
            pwa->next = wa->next;
            free(wa);
        }
    }

    free(node->as.dir.children);
    node->as.dir.children = NULL;
    node->as.dir.unfolded = false;
}

void lnk_free(struct Node* node) {
    free(node->as.link.readpath);
    if (node->as.link.to) node_free(node->as.link.to);
    free(node->as.link.to);
    node->as.link.readpath = NULL;
    node->as.link.to = NULL;
    node->as.link.tail = NULL;
}

void node_print(struct Node* node, struct Printer* pr) {
    pr->node(node);
    switch (node->type) {
        case Type_DIR: dir_print(node, pr); break;
        case Type_LNK: lnk_print(node, pr); break;
        default: ;
    }
}

void dir_print(struct Node* node, struct Printer* pr) {
    struct Dir dir = node->as.dir;
    if (dir.unfolded) {
        pr->enter(node);
        for (size_t k = 0; k < node->count; k++)
            node_print(dir.children[k], pr);
        pr->leave(node);
    }
}

void lnk_print(struct Node* node, struct Printer* pr) {
    if (node->as.link.tail && Type_DIR == node->as.link.tail->type)
        dir_print(node->as.link.tail, pr);
}

void lnk_resolve(struct Node* node) {
    char readpath[_MAX_PATH];
    ssize_t len = readlink(node->path, readpath, _MAX_PATH-1);
    if (len < 0) {
        may_strdup(node->as.link.readpath, strerror(errno));
        node->as.link.to = node->as.link.tail = NULL;
        return;
    }
    readpath[len] = '\0';

    char* savedpath; may_strdup(savedpath, readpath);

    char fullpath[_MAX_PATH];
    char* paste;
    char* copy;
    if ('/' == readpath[0]) {
        paste = copy = strcpy(fullpath, readpath)+1;
    } else {
        char* basedir = node->path;
        size_t lendir = strlen(basedir);
        while ('/' != basedir[lendir-1]) lendir--; // YYY: strrchr
        paste = memcpy(fullpath, basedir, lendir);
        paste+= lendir;
        copy = readpath;
    }

    if ('/' != paste[-1]) *paste++ = '/';
    do {
        if ('.' == *copy) {
            if ('.' == *(copy+1)) {
                paste--;
                if (fullpath == paste) {
                    node->as.link.to = node->as.link.tail = NULL;
                    return;
                }
                while ('/' != *--paste); // YYY: strrchr
                paste++;
            } else if ('/' == *(copy+1)) copy++;
            else *paste++ = *copy;
        } else if ('/' == *copy) {
            *paste++ = '/';
            while ('/' == *(copy+1)) copy++; // YYY: strchr
        } else *paste++ = *copy;
    } while (*copy++);
    paste--;
    while ('/' == *(paste-1)) paste--; // YYY: strrchr
    *paste = '\0';

    char* path; may_strdup(path, fullpath);

    node->as.link.readpath = savedpath;

    struct Node* niw = NULL;
    if (!(stat(node->path, &(struct stat){0}) < 0 && ELOOP == errno))
        niw = node_alloc(node->parent, path);

    node->as.link.to = niw;
    node->as.link.tail = niw && Type_LNK == niw->type
        ? niw->as.link.tail
        : niw;
}

static size_t _dir_new_child(struct Node* node, struct Node* true_parent, char* d_name) {
    size_t parent_path_len = strlen(node->path);

    if ('.' == d_name[0] && (
        '\0' == d_name[1]
        || ('.' == d_name[1] && '\0' == d_name[2])
        || !gflags.almost_all
    )) return -1;
    size_t nlen = strlen(d_name);
    if (gflags.ignore_backups && '~' == d_name[nlen-1]) return -1;

    size_t path_len = parent_path_len+2 + nlen;
    char* path; may_malloc(path, path_len);
    strcpy(path, node->path);

    char* name = path + parent_path_len;
    if ('/' != name[-1]) *name++ = '/';
    strcpy(name, d_name);

    struct Node* niw = node_alloc(true_parent, path);
    if (!niw) return -1;

    if (gflags.ignore) {
        bool ignore = selected_printer->filter
            ? selected_printer->filter(niw)
            : node_ignore(niw);
        if (ignore) {
            node_free(niw);
            free(niw);
            return -1;
        }
    }

    may_realloc(node->as.dir.children, (node->count+1) * sizeof(struct Node*));

    size_t k = 0;
    for (k = node->count; 0 < k; k--) {
        int cmp = node_compare(node->as.dir.children[k-1], niw, gflags.sort_order);
        if (!cmp) cmp = node_compare(node->as.dir.children[k-1], niw, Sort_NAME);
        if (cmp < 0) break;
        node->as.dir.children[k] = node->as.dir.children[k-1];
    }
    node->as.dir.children[k] = niw;

    node->count++;
    if (node != true_parent) true_parent->count++;
    return k;
}

void _dir_remove_child(struct Node* node, struct Node* true_parent, struct Node* child) {
    node->count--;
    if (node != true_parent) true_parent->count--;

    for (size_t k = child->index; k < node->count; k++) {
        node->as.dir.children[k] = node->as.dir.children[k+1];
        node->as.dir.children[k]->index--;
    }
    node->as.dir.children[node->count] = NULL;

    node_free(child);
    free(child);
    if (0 == node->count) {
        free(node->as.dir.children);
        node->as.dir.children = NULL;
    } else may_realloc(node->as.dir.children, node->count * sizeof(struct Node*));
}

void dir_unfold(struct Node* node) {
    struct Node* parent = node;
    if (Type_LNK == node->type) node = node->as.link.tail;
    if (!node || Type_DIR != node->type) return;

    node->as.dir.unfolded = true;
    if (node->as.dir.children) return;

    DIR *dir = opendir(node->path);
    if (dir) {
        struct dirent *ent;
        while ((ent = readdir(dir)))
            _dir_new_child(node, parent, ent->d_name);
        closedir(dir);
    }

    for (size_t k = 0; k < node->count; k++) {
        struct Node* child = node->as.dir.children[k];
        do {
            child->index = k;
        } while (Type_LNK == child->type && (child = child->as.link.to));
    }

    if (0 == (parent->count = node->count)) {
        free(node->as.dir.children);
        node->as.dir.children = NULL;
    } else may_realloc(node->as.dir.children, node->count * sizeof(struct Node*));

    if (gflags.watch) {
        uint32_t m = IN_ATTRIB | IN_CREATE | IN_DELETE | IN_MOVE;
        // YYY: should probably
        //if (0 == strcmp(root.path, fill.path)) m|= IN_DELETE_SELF | IN_MOVE_SELF;
        struct WatchAssoc* wa;
        may_malloc(wa, sizeof(struct WatchAssoc));
        wa->wd = inotify_add_watch(NOTIFY_FILENO, node->path, m);
        if (wa->wd < 0) die(node->path);
        wa->node = node;
        if (!watch_assocs) watch_assocs = wa;
        else watch_assocs->next = wa;
        wa->next = NULL;
    }
}

void dir_fold(struct Node* node) {
    if (Type_LNK == node->type) node = node->as.link.tail;
    if (!node || Type_DIR != node->type) return;

    if (!node->as.dir.children) dir_free(node);
    else node->as.dir.unfolded = false;
}

static void _recurse_dir_reload(struct Node* old, struct Node* niw) {
    if (Type_LNK == old->type) old = old->as.link.tail;
    if (!old || Type_DIR != old->type) return;

    if (Type_LNK == niw->type) niw = niw->as.link.tail;
    if (!niw || Type_DIR != niw->type) return;

    bool found_cursor = false;
    bool moved_cursor = false;

    // YYY: search could be better (maybe) by assuming little change in order
    //   - starting from max(old->index, niw->count-1)
    //   - if niw->count < old->count, go backward; else forward
    //   - reorder the conditions inside to have less strcmp avg case
    // from nlogn (always) to n (best) and n^2 (worst -- how often?)
    //
    // should it be 'optimized'? somewhat yes, because a use
    // case could be to 'watch' (ie. reload at interval)
    // hence reload should not be an intensive operation
    for (size_t i = 0; i < old->count; i++) {
        struct Node* old_child = old->as.dir.children[i];

        if (cursor == old_child) found_cursor = true;

        for (size_t j = 0; j < niw->count; j++) {
            struct Node* niw_child = niw->as.dir.children[j];

            if (0 == strcmp(old_child->name, niw_child->name)) {
                if (cursor == old_child) {
                    cursor = niw_child;
                    moved_cursor = true;
                }

                struct Node* d = Type_LNK == old_child->type
                    ? old_child->as.link.tail
                    : old_child;
                if (d && Type_DIR == d->type && d->as.dir.unfolded) {
                    dir_unfold(niw_child);
                    _recurse_dir_reload(old_child, niw_child);
                }

                break;
            } // if found
        } // for in niw
    } // for in old

    if (found_cursor && !moved_cursor) cursor = niw;
}
void dir_reload(struct Node* node) {
    struct Node* dir = node;
    if (Type_LNK == dir->type) dir = dir->as.link.tail;
    if (!dir || Type_DIR != dir->type) return;

    bool is_unfolded = dir->as.dir.unfolded;
    bool is_root = &root == node;

    if (is_root) {
        may_malloc(node, sizeof(struct Node));
        memcpy(node, &root, sizeof(struct Node));
    }

    char* niw_path; may_strdup(niw_path, node->path);
    struct Node* niw = node_alloc(node->parent, niw_path);

    if (is_root) {
        if (!niw) die("Cannot access root anymore");
        if (Type_DIR != niw->type) {
            errno = ENOTDIR;
            die(node->path);
        }
        memcpy(&root, niw, sizeof(struct Node));
        niw = &root;
    } else {
        if (!niw) {
            struct Node* parent = NULL;
            if (node->parent) {
                parent = Type_DIR == node->parent->type
                    ? node->parent
                    : node->parent->as.link.tail;
                parent = parent->as.dir.children[node->index];
            } else parent = node;

            _dir_remove_child(node->parent, parent, node);
            free(niw_path);
            return;
        } else node->parent->as.dir.children[niw->index = node->index] = niw;
    }

    if (is_unfolded) dir_unfold(niw);
    if (cursor == node) cursor = niw;
    _recurse_dir_reload(node, niw);
    node_free(node);
    free(node);
}

static struct termios orig_termios;
static bool atexit_set = false;

void term_restore(void) {
    if (!is_tty) return;

    if (tcsetattr(STDOUT_FILENO, TCSAFLUSH, &orig_termios) < 0) die("tcrstattr");
    is_raw = false;
}

void term_raw_mode(void) {
    if (!is_tty) return;

    if (!atexit_set) {
        if (tcgetattr(STDOUT_FILENO, &orig_termios) < 0) die("tcgetattr");
        if (0 != atexit(term_restore)) die("atexit");
    }

    // cfmakeraw: 1960 magic shit
    struct termios raw = orig_termios;
    raw.c_iflag &= ~(IGNBRK | BRKINT | PARMRK | ISTRIP | INLCR | IGNCR | ICRNL | IXON);
    raw.c_oflag &= ~(OPOST);
    raw.c_lflag &= ~(ECHO | ECHONL | ICANON | ISIG | IEXTEN);
    raw.c_cflag &= ~(CSIZE | PARENB);
    raw.c_cflag |= (CS8);

    if (tcsetattr(STDOUT_FILENO, TCSAFLUSH, &raw) < 0) die("tcsetattr");
    is_raw = true;
}

size_t ignore_count = 0;
char** ignore_list = NULL;
bool node_ignore(struct Node* node) {
    if (!ignore_list) return false;

    size_t cwd_len = strlen(cwd);
    if ('/' != memcmp(node->path, cwd, cwd_len+1)) return false;

    char* node_rel = node->path+strlen(cwd)+1;

    for (size_t k = 0; k < ignore_count; k++) {
        char* patt = ignore_list[k];
        size_t len = strlen(patt);

        bool invert = '!' == ignore_list[k][0];
        bool beginning = '/' == patt[0];
        bool middle = memchr(patt+1, '/', len-2);
        bool end = '/' == patt[len-1];

        if (end && Type_DIR != node->type) continue;
        if (invert || '\\' == ignore_list[k][0]) patt++;
        if (beginning) patt++;

        if (end) patt[len-1] = '\0'; // XXX hack
        int no = fnmatch(patt, beginning || middle ? node_rel : node->name, FNM_PATHNAME);
        if (end) patt[len-1] = '/'; // XXX hack

        if (!no) return true;
    }
    return false;
}

int node_compare(struct Node* node, struct Node* mate, enum Sort order) {
    if (Sort_REVERSE & order)
        return -node_compare(node, mate, order & ~Sort_REVERSE);

    if (Sort_DIRSFIRST & order) {
        bool node_dir = Type_DIR == (Type_LNK == node->type && node->as.link.tail ? node->as.link.tail : node)->type;
        bool mate_dir = Type_DIR == (Type_LNK == mate->type && mate->as.link.tail ? mate->as.link.tail : mate)->type;
        if (node_dir && !mate_dir) return -1;
        if (mate_dir && !node_dir) return +1;
        return node_compare(node, mate, order & ~Sort_DIRSFIRST);
    }

    switch (order) {
        case Sort_NAME: return strcoll(node->name, mate->name);

        case Sort_SIZE: return mate->stat.st_size - node->stat.st_size;

        case Sort_EXTENSION: {
            char* xa = strrchr(node->name, '.');
            if (!xa) return -1;
            char* xb = strrchr(mate->name, '.');
            if (!xb) return +1;
            return strcmp(xa+1, xb+1);
        }

        case Sort_ATIME: return mate->stat.st_atime - node->stat.st_atime;
        case Sort_MTIME: return mate->stat.st_mtime - node->stat.st_mtime;
        case Sort_CTIME: return mate->stat.st_ctime - node->stat.st_ctime;

        default: ;
    }
    return 0; // unreachable
}

bool user_was_stdin = false;
bool user_was_loopback = false;

static void _notify_events(void) {
    char buf[4096] __attribute__((aligned(__alignof__(struct inotify_event))));

    while (true) {
        ssize_t len = read(NOTIFY_FILENO, buf, sizeof(buf));
        if (len <= 0) return;

        struct inotify_event* event;
        for (char* head = buf; head < buf + len; head+= sizeof(struct inotify_event) + event->len) {
            event = (struct inotify_event *) head;

            //if (event->mask & (IN_DELETE_SELF | IN_MOVE_SELF))
            //    exit(EXIT_FAILURE); // only root node, exit

            if (!event->len) continue;

            struct Node* dir = NULL;
            struct WatchAssoc* wa = watch_assocs;
            do {
                if (event->wd == wa->wd) {
                    dir = wa->node;
                    break;
                }
            } while ((wa = wa->next));
            if (!dir) continue; // XXX: should not happen but could at least report it

            struct Node* parent = NULL;
            if (dir->parent) {
                parent = Type_DIR == dir->parent->type
                    ? dir->parent
                    : dir->parent->as.link.tail;
                parent = parent->as.dir.children[dir->index];
            } else parent = dir;

            if (event->mask & (IN_CREATE | IN_MOVED_TO)) {
                size_t index = _dir_new_child(dir, parent, event->name);
                // YYY: (3 lines) index fixup could be in _dir_new_child
                // at cost of a loss in efficiency in dir_unfold (2n -> nlogn)
                dir->as.dir.children[index]->index = index;
                while (++index < dir->count)
                    dir->as.dir.children[index]->index++;
                continue;
            }

            struct Node* file = NULL;
            for (size_t k = 0; k < dir->count; k++)
                if (0 == strcmp(event->name, dir->as.dir.children[k]->name))
                    file = dir->as.dir.children[k];
            if (!file) continue; // XXX: should not happend but could at least report it

            if (event->mask & (IN_DELETE | IN_MOVED_FROM)) {
                _dir_remove_child(dir, parent, file);
                continue;
            }

            if (event->mask & IN_ATTRIB) {
                // if (Type_REG == file->type || Type_EXEC == file->type) {
                    lstat(file->path, &file->stat);
                    file->type = S_IFMT & file->stat.st_mode;
                    if (file->stat.st_mode & S_IXUSR)
                        file->type = Type_EXEC;
                // }
                continue;
            }
        }
    }
}

int user_write(void* buf, size_t len) {
    return write(LOOPBACK_FILENO[LB_WRITE], buf, len);
}

int user_read(void* buf, size_t len) {
    fd_set cpy;
try_again: // jumps here when read retured 0 (EOF, closes fd)
    cpy = user_fds;
    if (select(9, &cpy, NULL, NULL, NULL) < 0) die("select");

    for (int i = 8; -1 < i; i--)
        if (FD_ISSET(i, &cpy)) {
            if (gflags.watch && NOTIFY_FILENO == i) {
                _notify_events();
                user_was_stdin = false;
                user_was_loopback = true;
                char* d = (char*)buf;
                *d = CTRL('L');
                return 1;
            }

            user_was_stdin = STDIN_FILENO == i;
            user_was_loopback = LOOPBACK_FILENO[LB_READ] == i;
            ssize_t r = read(i, buf, len);
            if (0 == r) {
                FD_CLR(i, &user_fds);
                close(i);
                goto try_again;
            }
            return r;
        }

    return -1; // unreachable
}

void watch_start(void) {
    NOTIFY_FILENO = inotify_init1(IN_NONBLOCK);
    if (NOTIFY_FILENO < 0) die("notify");
    FD_SET(NOTIFY_FILENO, &user_fds);
}
void watch_stop(void) {
    if (watch_assocs) {
        struct WatchAssoc* wa;
        struct WatchAssoc* pwa = watch_assocs;
        do {
            wa = pwa->next;
            //inotify_rm_watch(NOTIFY_FILENO, pwa->wd);
            free(pwa);
        } while ((pwa = wa));
        close(NOTIFY_FILENO);
        FD_CLR(NOTIFY_FILENO, &user_fds);
        watch_assocs = NULL;
    }
}

static char** printer_argv = NULL;
static int printer_argc = 0;
static char* rcfile = NULL;
char* opts(int argc, char* argv[]) {
    selected_printer = &ascii_printer;
    char* selected_path = NULL;

    for (int k = 0; k < argc; k++) {
        if (0 == strcmp("--help", argv[k])) {
            printf("Usage: %s [--printer=NAME] [--LONGOPTIONS] [-FLAGS] [[--] ROOT]\n", prog);
            exit(EXIT_FAILURE);
        } else if (0 == strcmp("--version", argv[k])) {
            puts(
                TREEST_VERSION
                #ifdef FEAT_READLINE
                "\n+ readline"
                #endif
                #ifdef FEAT_GIT2
                "\n+ git2"
                #endif
            );
            exit(EXIT_SUCCESS);
        } else if (0 == memcmp("--printer=", argv[k], 10)) {
            char* arg = argv[k] + 10;
            #define DO(it) if (0 == strcmp(it.name, arg)) selected_printer = &it;
            EVERY_PRINTERS(DO, else)
            #undef DO
            else {
                printf("No such printer: '%s'\n", arg);
                exit(EXIT_FAILURE);
            }
        } else if (0 == memcmp("--ignore=", argv[k], 9)) {
            char* arg = argv[k] + 9;
            ignore_count++;
            may_realloc(ignore_list, ignore_count * sizeof(char*));
            ignore_list[ignore_count-1] = arg;
        } else if (0 == memcmp("--rcfile=", argv[k], 9)) {
            rcfile = argv[k] + 9;
        } else {
            if ('-' == argv[k][0]) {
                if ('-' == argv[k][1] && '\0' == argv[k][2]) {
                    selected_path = argv[k+1];
                    break;
                }
                printer_argc++;
                may_realloc(printer_argv, printer_argc * sizeof(char*));
                printer_argv[printer_argc-1] = argv[k];
            } else {
                selected_path = argv[k];
                break;
            }
        } // else (argv not long option)
    } // foreach argv

    return selected_path;
}

int main(int argc, char* argv[]) {
    #ifdef TRACE_ALLOCS
    mtrace();
    #endif
    if (!getcwd(cwd, _MAX_PATH)) die("getcwd");
    setlocale(LC_ALL, "");

    prog = argv[0];
    argv++;
    argc--;

    char* arg_path = opts(argc, argv);
    if (!arg_path) arg_path = cwd;
    char* path;
    if (!(path = realpath(arg_path, NULL))) die(arg_path);

    if (lstat(path, &root.stat) < 0) die(path);
    if (!S_ISDIR(root.stat.st_mode)) {
        errno = ENOTDIR;
        die(path);
    }
    root.path = path;
    root.name = strrchr(path, '/')+1;
    root.type = Type_DIR;

    cursor = &root;

    if (pipe(LOOPBACK_FILENO) < 0) die("pipe(loopback)");
    // YYY: should probably O_NONBLOCK, but the real hack
    // is to fork exec a while (read && write); in the
    // mean time, no writing more than PIPE_BUF without
    // causing a deadlock - not that it matter 'cause this
    // 'feature' should not be used anyway...
    FD_ZERO(&user_fds);
    FD_SET(STDIN_FILENO, &user_fds);
    FD_SET(LOOPBACK_FILENO[LB_READ], &user_fds);

    if (selected_printer->init) selected_printer->init();
    for (int k = 0; k < printer_argc; k++) {
        if ('-' == printer_argv[k][1]) {
            if (!selected_printer->command || !selected_printer->command(printer_argv[k]+2)) {
                printf("Unknown command for '%s': '%s'\n", selected_printer->name, printer_argv[k]+2);
                if (selected_printer->del) selected_printer->del();
                exit(EXIT_FAILURE);
            }
            continue;
        }
        char* flag = printer_argv[k];
        while (*++flag)
            selected_printer->toggle
                ? selected_printer->toggle(*flag)
                : toggle_gflag(*flag);
    }
    free(printer_argv);
    printer_argv = NULL;

    dir_unfold(&root);

    if (rcfile) {
        int rcfd = open(rcfile, O_RDONLY);
        if (rcfd < 0) die(rcfile);
        FD_SET(rcfd, &user_fds);
    }

    if ((is_tty = isatty(STDOUT_FILENO))) term_raw_mode();

    char user;
    while (true) {
        if (command_map[CTRL('L')].f)
            command_map[CTRL('L')].f();

        do {
            int r = user_read(&user, 1);
            if (r < 0) die("read");
            if (0 == r && user_was_stdin)
                return EXIT_SUCCESS;

            if ('.' != user) {
                may_realloc(register_map['.'], 2 * sizeof(char));
                register_map['.'][0] = user;
                register_map['.'][1] = '\0';
            }
        } while (!run_command(user));
    }
}
